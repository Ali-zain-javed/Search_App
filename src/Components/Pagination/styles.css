

.data {
  float: left;
  color: grey;
  font-weight: 600;
}
.pagination {
  display: inline-block;
  float: right;
}

.pagination a {
  color: black;
  float: left;
  padding: 8px 16px;
  text-decoration: none;
  transition: background-color .3s;
  border: 1px solid #ddd;
  margin: 0 4px;
}

.pagination a.active {
  background-color: #04AA6D;
  color: white;
  border: 1px solid #04AA6D;
}

.pagination a:hover:not(.active) {background-color: #ddd;}