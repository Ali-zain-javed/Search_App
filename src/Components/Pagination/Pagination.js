import React, { useState } from 'react';
import './styles.css';
const LEFT_PAGE = 'LEFT';
const RIGHT_PAGE = 'RIGHT';

/**
 * Helper method for creating a range of numbers
 * range(1, 5) => [1, 2, 3, 4, 5]
 */
const range = (from, to, step = 1) => {
  let i = from;
  const rangeArray = [];

  while (i <= to) {
    rangeArray.push(i);
    i += step;
  }

  return rangeArray;
};
let totalPages = 0;
const Pagination = ({
  totalRecords = 0,
  pageLimit = 10,
  pageNeighbors = 3,
  currentPage = 1,
  onPageChanged,
}) => {
  const [state, setState] = useState({
    currentPage,
    pageLimit,
    totalPages,
    totalRecords,
  });

  React.useEffect(() => {
    // pageNeighbors can be: 0, 1 or 2
    pageNeighbors = Math.max(0, Math.min(pageNeighbors, 2));
    totalPages = Math.ceil(totalRecords / pageLimit);

    setState({
      ...state,
      currentPage,
      pageLimit,
      totalPages,
      totalRecords,
    });
  }, [totalRecords, pageLimit, currentPage]);

  const gotoPage = (page) => {
    const paginationData = {
      currentPage: Math.max(0, Math.min(page, totalPages)),
      pageLimit,
      totalPages,
      totalRecords,
    };

    onPageChanged(paginationData.currentPage);
    // setState(paginationData);
  };

  const handleClick = (page) => (evt) => {
    evt.preventDefault();
    gotoPage(page);
  };

  const handleMoveLeft = (evt) => {
    evt.preventDefault();
    // gotoPage(state.currentPage - pageNeighbors * 2 - 1);
    gotoPage(state.currentPage - 1);
  };

  const handleMoveRight = (evt) => {
    evt.preventDefault();
    // gotoPage(state.currentPage + pageNeighbors * 2 + 1);
    gotoPage(state.currentPage + 1);
  };

  /**
   * Let's say we have 10 pages and we set pageNeighbors to 2
   * Given that the current page is 6
   * The pagination control will look like the following:
   *
   * (1) < {4 5} [6] {7 8} > (10)
   *
   * (x) => terminal pages: first and last page(always visible)
   * [x] => represents current page
   * {...x} => represents page neighbors
   */

  const fetchPageNumbers = () => {
    const { currentPage } = state;

    /**
     * totalNumbers: the total page numbers to show on the control
     * totalBlocks: totalNumbers + 2 to cover for the left(<) and right(>) controls
     */
    const totalNumbers = pageNeighbors * 2 + 3;
    const totalBlocks = totalNumbers + 2;

    if (totalPages > totalBlocks) {
      const startPage = Math.max(2, currentPage - pageNeighbors);
      const endPage = Math.min(totalPages - 1, currentPage + pageNeighbors);

      let pages = range(startPage, endPage);

      /**
       * hasLeftSpill: has hidden pages to the left
       * hasRightSpill: has hidden pages to the right
       * spillOffset: number of hidden pages either to the left or to the right
       */
      const hasLeftSpill = startPage > 2;
      const hasRightSpill = totalPages - endPage > 1;
      const spillOffset = totalNumbers - (pages.length + 1);

      switch (true) {
        // handle: (1) < {5 6} [7] {8 9} (10)
        case hasLeftSpill && !hasRightSpill: {
          const extraPages = range(startPage - spillOffset, startPage - 1);
          pages = [LEFT_PAGE, ...extraPages, ...pages];
          break;
        }

        // handle: (1) {2 3} [4] {5 6} > (10)
        case !hasLeftSpill && hasRightSpill: {
          const extraPages = range(endPage + 1, endPage + spillOffset);
          pages = [...pages, ...extraPages, RIGHT_PAGE];
          break;
        }

        // handle: (1) < {4 5} [6] {7 8} > (10)
        case hasLeftSpill && hasRightSpill:
        default: {
          pages = [LEFT_PAGE, ...pages, RIGHT_PAGE];
          break;
        }
      }
      return [1, ...pages, totalPages];
    }
    return range(1, totalPages);
  };

  if (!totalRecords || totalPages === 1) {
    return null;
  }

  const allPages = fetchPageNumbers();

  return (
    <div>
      <div className="data">
        Showing {currentPage * pageLimit - pageLimit + 1} - {currentPage * pageLimit} of {totalRecords} items
      </div>
      <div className="pagination">
        {allPages.map((page, index) => {
          if (page === LEFT_PAGE) {
            return (<a href="#" key={index}  onClick={handleMoveLeft}>&laquo;</a>)
          }
          if (page === RIGHT_PAGE) {
            return (
              <a href="#" key={index}  onClick={handleMoveRight}>&raquo;</a>
            )
          }
          return (
            <a href="#" key={index} className={state.currentPage === page ? "active" : ""} onClick={state.currentPage === page ? () => { } : handleClick(page)}>{page}</a>)
        })}
      </div>
    </div>
  );
};

export default Pagination;
