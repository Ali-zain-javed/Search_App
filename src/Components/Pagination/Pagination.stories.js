import React from 'react';
import Pagination from './Pagination.js'
export default {
  component: Pagination,
  title: 'Components/Pagination',
};

//👇 We create a “template” of how args map to rendering
const Template = (args) => <Pagination {...args} />;

//👇 Each story then reuses that template
export const Primary = Template.bind({});

Primary.args = {
  currentPage: 1,
  pageLimit: 10,
  totalRecords: 100, onPageChanged: (pageNumber) => { console.log(pageNumber) }
};