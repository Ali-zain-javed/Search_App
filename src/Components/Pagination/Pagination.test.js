import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import React from 'react';
import { configure, mount, render, shallow } from 'enzyme';
import Pagination from './Pagination.js';
configure({ adapter: new Adapter() });
import ReactDOM from 'react-dom';
const matchSnapshot = (event) => { };

//moke data for testing
const props = {
    currentPage: 1,
    pageLimit: 10,
    totalRecords: 100, onPageChanged: () => { }
}
describe('MyComponent', () => {
    it('In my test case should render correctly and save snapshots', () => {
        const component = shallow(
            <Pagination {...props} />
        );
        expect(component).toMatchSnapshot();
    });
});
it('In my test case should run pagination component without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
        <Pagination {...props} />
        ,
        div
    );
    ReactDOM.unmountComponentAtNode(div);
});
it('In my test case should run pagination component with page 2 without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
        <Pagination {...props} currentPage={2} />
        ,
        div
    );
    ReactDOM.unmountComponentAtNode(div);
});

it('In my test case my form should exist label field', () => {
    const wrapper = mount(
        <Pagination {...props} currentPage={2} totalRecords={200}/>
    );
    expect(wrapper.text()).toContain('Showing');
    expect(wrapper.text()).toContain('1');
    expect(wrapper.text()).toContain('2');
    expect(wrapper.text()).toContain('items');
    expect(wrapper.text()).toContain('200');
});

