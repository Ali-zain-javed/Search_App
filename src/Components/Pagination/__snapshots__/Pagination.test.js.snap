// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent In my test case should render correctly and save snapshots 1`] = `ShallowWrapper {}`;
